### PR TITLE
[WIP] Enable reproducibility for distributed trainings

### PR DESCRIPTION
**Pull request by hasansalimkanmaz**
# What does this PR do?
This PR ensures reproducibility for distributed trainings by setting seed for worker in dataloader and setting environment variables for cuda.

This PR is motivated by [this issue](https://github.com/huggingface/transformers/issues/16549#). 

## Who can review?
@saattrupdan @sgugger I am looking forward to your feedback

### CODE REVIEW

File: src/transformers/trainer_utils.py [hunk 1]
```diff
@@ -46,7 +46,27 @@
     import tensorflow as tf
 
 
-def set_seed(seed: int):
+def seed_worker():
+    worker_seed = torch.initial_seed() % 2**32
+    set_seed(worker_seed)
+
+
+def set_deterministic_seed_for_cuda():
```
Comment (saattrupdan) [inline src/transformers/trainer_utils.py:54]: I'd change the name of this function, as it doesn't set a seed at all. Perhaps `enable_determinism`?

File: src/transformers/trainer_utils.py [hunk 2]
```diff
@@ -46,7 +46,27 @@
...
+    #  Enable PyTorch deterministic mode. This potentially requires either the environment
```
Comment (saattrupdan) [inline src/transformers/trainer_utils.py:55]: Function missing docstring. This docstring could also contain a link to the [reproducibility page of the PyTorch docs](https://pytorch.org/docs/stable/notes/randomness.html).

File: src/transformers/trainer_utils.py [hunk 3]
```diff
@@ -46,7 +46,27 @@
     import tensorflow as tf
 
 
-def set_seed(seed: int):
+def seed_worker():
+    worker_seed = torch.initial_seed() % 2**32
```
Comment (saattrupdan) [inline src/transformers/trainer_utils.py:50]: Function missing docstring.

File: src/transformers/trainer_utils.py [hunk 4]
```diff
@@ -46,7 +46,27 @@
...
+    set_seed(worker_seed)
+
+
+def set_deterministic_seed_for_cuda():
+    #  Enable PyTorch deterministic mode. This potentially requires either the environment
+    #  variable 'CUDA_LAUNCH_BLOCKING' or 'CUBLAS_WORKSPACE_CONFIG' to be set,
+    # depending on the CUDA version, so we set them both here
+    os.environ["CUDA_LAUNCH_BLOCKING"] = "1"
+    os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":16:8"
```
Comment (saattrupdan) [inline src/transformers/trainer_utils.py:59]: These are only relevant if torch is available, so I'd move them down into the if-block.

File: src/transformers/trainer_utils.py [hunk 5]
```diff
@@ -46,7 +46,27 @@
...
+
+    if is_torch_available():
+        torch.use_deterministic_algorithms(True)
+
+        # Enable CUDNN deterministic mode
+        torch.backends.cudnn.deterministic = True
+        torch.backends.cudnn.benchmark = False
+
+
+def set_seed(seed: int, set_seed_for_cuda: bool = True):
```
Comment (saattrupdan) [inline src/transformers/trainer_utils.py:69]: Related to the function name above, I'd argue that the argument here should be changed to something like `enable_determinism`. Further, I'd make the default `False`, as enabling it _can_ cause weird errors, if one uses algorithms that don't have a deterministic variant yet.

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -704,6 +705,7 @@ def get_train_dataloader(self) -> DataLoader:
             drop_last=self.args.dataloader_drop_last,
             num_workers=self.args.dataloader_num_workers,
             pin_memory=self.args.dataloader_pin_memory,
+            worker_init_fn=seed_worker,
```
Comment (saattrupdan) [inline src/transformers/trainer.py:708]: Is this something that should always be enabled, or should it check if `self.args.seed` is set and only set it in case it's not `None`?

**Review by saattrupdan — CHANGES_REQUESTED** [review]
Thanks for investing your time to implement this PR! 😊 

I have mostly small changes related to documentation and naming, but otherwise looks good 👍 

EDIT: To enable support for Tensorflow models, you could add use the [enable_op_determinism](https://www.tensorflow.org/api_docs/python/tf/config/experimental/enable_op_determinism) in the Tensorflow case.

File: src/transformers/trainer_utils.py [hunk 7]
```diff
@@ -46,7 +46,38 @@
...
+
+        # Enable CUDNN deterministic mode
+        torch.backends.cudnn.deterministic = True
+        torch.backends.cudnn.benchmark = False
+
+    if is_tf_available():
+        tf.config.experimental.enable_op_determinism()
+
+
+def set_seed(seed: int, enable_determinism: bool = True):
```
Comment (sgugger) [inline src/transformers/trainer_utils.py:80]:
```suggestion
def set_seed(seed: int, distributed_determinism: bool = False):
```
I like `distribitude_determinism` a bit better. Since this is a new addition, the default should be set to `False`. Although it does fix what one might consider a bug, so I'm not sure on this one. @LysandreJik do you have an opinion?

File: src/transformers/trainer.py (hunk 6 above)
Comment (sgugger) [reply src/transformers/trainer.py:708]: I think it doesn't hurt doing it all the time.

**Review by sgugger — APPROVED** [review]
Thanks for your work on this!

**Review by LysandreJik — COMMENTED** [review]
Thanks for working on this, that's an important feature! So as to not introduce a breaking change, and for clarity of the API, I'd personally vouch for not adding the `enable_determinism` flag to the `set_seed` method.

From the title of the method I understand it should set the seed, and that's it. I don't think it should do anything else. However, the `enable_determinism_for_distributed_training` method likely needs the seed to be set in order to benefit from full determinism, so I'd even push to have the `set_seed` method called inside the `enable_determinism_for_distributed_training`, adding a `seed` argument to that last method.

What do you think?

**Comment by hasansalimkanmaz** [comment]
> 

I like this idea. I can implement it after we reach a conclusion on it, however, it is not clear to me how to implement it. Could you point me to which parts of the code I need to change/pay attention not to break anything if we decide to go for this idea?

File: src/transformers/trainer_utils.py [hunk 8]
```diff
@@ -46,7 +46,38 @@
...
-def set_seed(seed: int):
+def seed_worker():
+    """
+    Helper function to set worker seed during Dataloader initialization.
+    """
+    worker_seed = torch.initial_seed() % 2**32
+    set_seed(worker_seed)
+
+
+def enable_determinism_for_distributed_training():
```
Comment (sgugger) [inline src/transformers/trainer_utils.py:57]: The idea would be for this function to take `seed` here.
```suggestion
def enable_full_determinism(seed: int):
```
and then call `set_seed` inside (instead of `set_seed` calling this function).

(Also changing the name to be a bit shorter.)

File: src/transformers/trainer_utils.py [hunk 9]
```diff
@@ -61,6 +92,8 @@ def set_seed(seed: int):
         # ^^ safe to call this function even if cuda is not available
     if is_tf_available():
         tf.random.set_seed(seed)
+    if enable_determinism:
+        enable_determinism_for_distributed_training()
```
Comment (sgugger) [inline src/transformers/trainer_utils.py:96]: And so this part would disappear here, it would be the other way around.

**Review by sgugger — COMMENTED** [review]
Here are some pointer on what @LysandreJik suggests.

**Comment by hasansalimkanmaz** [comment]
@sgugger Thanks for the pointers and sorry for not being so clear. I would like to know in which places `enable_full_determinism` should be called. Currently, `set_seed` is called several places in the codebase. I don't think these calls will be replaced with `enable_full_determinism`. 

With the latest commits, I have already addressed your pointers. Now I am waiting your feedback for where to call `enable_full_determinism` in the codebase. It is not called any place in the codebase right now.

**Comment by sgugger** [comment]
There can be an added flag in the `TrainingArguments` and we can call this function instead of `set_seed` in the `Trainer`. Otherwise it will be for the users to use this one instead of `set_seed` in their own scripts (you should make it accessible in the main init by the way!)

**Comment by hasansalimkanmaz** [comment]
@sgugger I think I have addressed all your comments. Is there anything that needs to be done for this PR?

File: src/transformers/training_args.py [hunk 10]
```diff
@@ -446,6 +446,9 @@ class TrainingArguments:
         auto_find_batch_size (`bool`, *optional*, defaults to `False`)
             Whether to find a batch size that will fit into memory automatically through exponential decay, avoiding
             CUDA Out-of-Memory errors. Requires accelerate to be installed (`pip install accelerate`)
+        full_determinism (`bool`, *optional*, defaults to `False`)
+            If True, enable_full_determinism is called instead of set_seed to ensure reproducible results in
```
Comment (sgugger) [inline src/transformers/training_args.py:450]:
```suggestion
            If `True`, [`enable_full_determinism`] is called instead of [`set_seed`] to ensure reproducible results in
```

**Review by sgugger — APPROVED** [review]
Thanks, just one last comment on the doc!

**Comment by hasansalimkanmaz** [comment]
Is it normal that 3 tests fail suddenly after a commit in a docstring? I couldn't understand why tests are failing.

**Comment by sgugger** [comment]
Those are just flaky, no link to your PR. Thanks again for all your work on this!

**Comment by alexcoca** [comment]
@sgugger @hasansalimkanmaz I had a question about this PR - why is it necessary to set `CUDA_LAUNCH_BLOCKING`? This disables asynchronous execution of CUDA programs, but the cuda/pytorch docs don't mention it necessary for deterministic training?

**Comment by saattrupdan** [comment]
@alexcoca It's required to make some CUDA algorithms deterministic if the CUDA version is older than 10.2. I suppose it could be replaced by a CUDA version check somehow, and only using it if it's an old version?

**Comment by alexcoca** [comment]
@saattrupdan I would go for this approach, because running the CUDA programs in asynchronous mode will definitely slow things down beyond belief. I implemented this PR myself without the `CUDA_LAUNCH_BLOCKING` setting and will report if I manage to preserve determinism.

**Comment by alexcoca** [comment]
I experimented with training a dialogue state tracking model on the SGD corpus starting from Google's v1.1 T5 (220M) paramaters. I allowed the model to train for roughly two epoch and evaluated task oriented performance every 2k steps (max train steps was 12k). 

Ran 4 experiments: 2 in which I set the seed, and an additional 2 where I do roughly the same as `ensure_determinism` except setting `CUDA_LAUNCH_BLOCKING`. I also set `CUBLAS_WORKSPACE_CONFIG=':4096:8'`. Each experiment was trained on 2 A100-80GB with `cuda/11.4 openmpi/4.1.1/gcc-9.4.0-epagguv`, `pytorch 1.10` and transformers `4.19.2`. You can see below that I was able to reproduce the metrics in all runs and with no major performance hits. I guess that convolution benchmarking and non-det ops are less relevant for pytorch. With 4.17.2 the performance was wreaking havoc on the same seed, sign that the data ordering was the culprit. 

image

**Comment by sgugger** [comment]
Agreed for the first one. For the second one, we could avoid overriding an existing `CUBLAS_WORKSPACE_CONFIG` if it's already in the env? In all cases, it should be clearly stated in the doc of the flag that triggers the full reproducibility that it comes at a performance price.

**Comment by alexcoca** [comment]
Yes, I agree with the above! I'm at ACL next week but I'll try and open a small PR to address this the week after!

**Comment by hasansalimkanmaz** [comment]
Thanks, @alexcoca for noticing this and for your time.
